File: 301_400/378_ Kth_Smallest_Element_in_a_Sorted_Matrix/Solution.py
class Solution:
    def kthSmallest(self, matrix: List[List[int]], k: int) -> int:
        row, col = len(matrix),len(matrix[0])
        left, right = matrix[0][0], matrix[row-1][col-1] 
        while left < right:
            mid = (left + right) >> 1
            if self.lessThanMid(matrix, mid) < k:
                left = mid + 1
            else:
                right = mid
        return left
    
    def lessThanMid(self, matrix, mid):
        r, c = len(matrix), len(matrix[0]) 
        i, j, count = r - 1, 0, 0
        while i >= 0 and j < c:
            if matrix[i][j] <= mid:
                count += i + 1
                j += 1
            else:
                i -= 1
        return count

File: 301_400/378_ Kth_Smallest_Element_in_a_Sorted_Matrix/Solution.rs
impl Solution {
    pub fn kth_smallest(matrix: Vec<Vec<i32>>, k: i32) -> i32 {
        let (row, col) = (matrix.len(), matrix[0].len());
        let (mut l,mut r) = (matrix[0][0], matrix[row-1][col-1]);
        while l < r {
            let mid = (l + r) >> 1;
            if Solution::less_than_mid(&matrix, mid) < k {
                l = mid + 1; 
            } else {
                r = mid;
            }
        }
        l
    }
    pub fn less_than_mid(m: &Vec<Vec<i32>>, mid: i32) -> i32 {
        let (row, col) = (m.len(), m[0].len());
        let (mut i, mut j,mut count) = ((row - 1) as i32, 0_usize, 0_i32);
        while i >= 0 && j < col {
            if m[i as usize][j] <= mid {
                count += (i + 1);
                j += 1;
            } else {
                i -= 1; 
            }
        }
        count
    } 
}
